File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.py
"""
Time = O(n)
Space = O(n)
"""
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        num_set = set(nums)
        longest = 0
        
        # Traverse throught the array
        # O(n)
        for i in range(len(nums)):
            if nums[i] - 1 in num_set:
                continue
            # Found the first number in a certain sequence
            cur_length = 1
            cur_num = nums[i]
            while (cur_num + 1) in num_set:
                cur_length += 1
                cur_num += 1
            
            # Update the longest
            longest = max(longest, cur_length)
        
        return longest
        

File: 0128-longest-consecutive-sequence/0128-longest-consecutive-sequence.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        int longestLength = 0, curLength = 0, curNum;
        unordered_set<int> numSet;

        for (int num: nums) {
            numSet.insert(num);
        }

        for (int i = 0; i < nums.size(); i++) {
            if (numSet.find(nums[i] - 1) != numSet.end()) {
                continue;
            }

            curLength = 1;
            curNum = nums[i];

            while (numSet.find(curNum + 1) != numSet.end()) {
                curLength++;
                curNum++;
            }

            longestLength = max(longestLength, curLength); 
        }

        return longestLength;
    }
};